File: shatter-go/instrument/entropy.go
// Package instrument provides runtime instrumentation for the Go frontend.

package instrument

import "math"

// MinBufferSize is the minimum byte count for reliable entropy measurement.
const MinBufferSize = 32

// EntropyDeltaThreshold is the minimum bits/byte difference between input
// and output entropy to classify as likely crypto.
const EntropyDeltaThreshold = 1.5

// ShannonEntropy computes Shannon entropy in bits per byte (0.0 to 8.0).
// Returns 0.0 for slices shorter than MinBufferSize.
func ShannonEntropy(data []byte) float64 {
	if len(data) < MinBufferSize {
		return 0.0
	}

	var counts [256]uint64
	for _, b := range data {
		counts[b]++
	}

	length := float64(len(data))
	entropy := 0.0
	for _, count := range counts {
		if count > 0 {
			p := float64(count) / length
			entropy -= p * math.Log2(p)
		}
	}

	return entropy
}

File: shatter-ts/src/entropy.ts
/**
 * Shannon entropy measurement for runtime crypto detection.
 *
 * Measures byte-level entropy of buffers to confirm cryptographic behavior:
 * encryption produces high-entropy output from low-entropy input,
 * decryption does the reverse.
 */

/** Buffers smaller than this are too short for reliable entropy measurement. */
export const MIN_BUFFER_SIZE = 32;

/** Minimum bits/byte difference to classify as likely crypto. */
export const ENTROPY_DELTA_THRESHOLD = 1.5;

/**
 * Compute Shannon entropy in bits per byte (0.0 to 8.0).
 *
 * Returns 0.0 for empty buffers or those shorter than {@link MIN_BUFFER_SIZE}.
 * For uniform random data, approaches 8.0; for constant data, returns 0.0.
 */
export function shannonEntropy(data: Buffer): number {
  if (data.length < MIN_BUFFER_SIZE) {
    return 0.0;
  }

  const counts = new Float64Array(256);
  for (let i = 0; i < data.length; i++) {
    const byte = data[i] as number;
    counts[byte] = (counts[byte] as number) + 1;
  }

  const len = data.length;
  let entropy = 0.0;
  for (let i = 0; i < 256; i++) {
    const count = counts[i] as number;
    if (count > 0) {
      const p = count / len;
      entropy -= p * Math.log2(p);
    }
  }

  return entropy;
}
